File: AIP/src/framework/AStar.java
package framework;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by Olav on 27/08/2014.
 */
public class AStar<T extends AStarState> implements Runnable {

    public final static int TYPE_BEST_FIRST = 0;
    public final static int TYPE_BREADTH_FIRST = 1;
    public final static int TYPE_DEPTH_FIRST = 2;

    private long stepInterval = 0;
    private long lastStepTime = 0;

    AStarStateHandler handler;
    T initialState;
    T goalState;

    public AStarNode<T> initialNode = null;
    public AStarNode<T> goalNode = null;
    public AStarNode<T> currentNode = null;

    Callback<AStar> callback;

    private int type;


    Map<Integer, AStarNode<T>> nodes;
    public List<AStarNode<T>> open;

    public int expandedNodes = 0;
    public boolean success = false;
    /**
     * @param stateHandler
     * @param initialState
     * @param goalState
     * @param callback - This callback will be called right after each node is popped off open.
     */
    public AStar(AStarStateHandler stateHandler, T initialState, T goalState, Callback<AStar> callback) {

        this.handler = stateHandler;
        this.callback = callback;

        setNewStates(initialState, goalState);

        initialize();
    }

    public AStar(AStarStateHandler stateHandler, T initialState, T goalState) {
        this(stateHandler, initialState, goalState, null);
    }

    public void setNewStates(T initialState, T goalState) {
        initialState.setHash(handler.generateHash(initialState));
        initialState.setH(handler.calculateH(initialState, goalState));
        this.initialState = initialState;

        goalState.setHash(handler.generateHash(goalState));
        this.goalState = goalState;
    }

        /**
         * This method initializes all datastructures required to run the algorithm.
         * This method is automagically called from the constructor.
         */
    public void initialize() {
        this.open = new ArrayList<AStarNode<T>>();
        this.nodes = new HashMap<Integer, AStarNode<T>>();
        currentNode = null;
        this.expandedNodes = 0;

        this.initialNode = new AStarNode(initialState);
        this.goalNode = new AStarNode(goalState);

        nodes.put(initialNode.getHash(), initialNode);
        open.add(initialNode);
    }

    public void setStepInterval(long ms) {
        this.stepInterval = ms;
    }

    public void setCallback(Callback<AStar> callback) {
        this.callback = callback;
    }

    /**
     * Sets the searchtype.
     * @param searchType
     * @return
     */
    public AStar<T> setType(int searchType) {
        this.type = searchType;
        return this;
    }

    @Override
    public void run() {
        success = false;

        do {
            //Check if we can step.
            long now = System.currentTimeMillis();
            if (now - lastStepTime < stepInterval) {
                continue;
            }
            lastStepTime = now;

            this.currentNode = open.remove(0);
            if (callback != null) {
                callback.callback(this);
            }
            //Check success criteria
            if (currentNode.equals(goalNode)){
                goalNode = currentNode;
                success = true;
                break;
            }

            //Expand
            List<T> children = handler.getChildren(currentNode.state);
            currentNode.isExpanded = true;
            expandedNodes++;

            for(T child : children) {
                AStarNode childNode;
                if(nodes.containsKey(child.getHash())) {
                    childNode = nodes.get(child.getHash());
                }
                else {
                    childNode = new AStarNode(child);
                }

                //Add childNodeToParent and set the distance between them.
                currentNode.addChild(childNode, childNode.state.getG() - currentNode.state.getG());

                if(!nodes.containsKey(childNode.getHash())) {
                    attachAndEval(currentNode, childNode);
                    nodes.put(childNode.getHash(), childNode);
                    insert(childNode);
                }
                //Node was already made, check if new path is more optimal than previous path found
                else if (childNode.state.getG() > currentNode.state.getG() + currentNode.getArcCost(childNode)) {
                    attachAndEval(currentNode, childNode);

                    //See if propagation to grandchildren is needed.
                    if(childNode.isExpanded){
                        propagate(childNode);
                    }
                    //If not, we need to put it in the correct place in the open list.
                    else {
                        open.remove(childNode);
                        insert(childNode);
                    }
                }
            }

        } while(open.size() > 0);
        callback.callback(this);
        System.out.println("Nodes expanded: " + expandedNodes);
    }

    public void insert(AStarNode<T> node) {
        switch(type) {
            case TYPE_DEPTH_FIRST:
                open.add(0, node);
                break;
            case TYPE_BREADTH_FIRST:
                open.add(node);
                break;
            case TYPE_BEST_FIRST:
                open.add(binarySearch(open, 0, open.size(), node.getF()), node);
                break;
            default:
                throw new IllegalStateException("Searchtype is invalid");
        }
    }

    private void propagate(AStarNode<T> parentNode) {
        double g = parentNode.state.getG();
        for (AStarNode child : parentNode.getChildren()) {
            double newG = g + parentNode.getArcCost(child);
            if (newG < child.state.getG()) {
                child.setParent(parentNode);
                child.state.setG(g);
                if(!child.isExpanded) {
                    open.remove(child);
                    insert(child);
                }
                propagate(child);
            }
        }
    }

    private void attachAndEval(AStarNode parentNode, AStarNode childNode) {
        childNode.setParent(parentNode);
        childNode.state.setH(handler.calculateH(childNode.state, goalNode.state));
    }

    private int binarySearch(List<AStarNode<T>> a, int start, int end, double val){

        if(start == end){
            return start;
        }
        int middle = (start + end)/2;
        if(val <= a.get(middle).getF()){
            return binarySearch(a, start, middle, val);
        }
        else{
            return binarySearch(a, middle + 1, end, val);
        }
    }
}

File: AIP/py/AStar.py
from ASNode import ASNode
import time


MODE_BEST_FIRST = 0
MODE_DEPTH_FIRST = 1
MODE_BREADTH_FIRST = 2

class AStar(object):
	
	def __init__(self, renderer = None):
		self.OPEN = None
		self.nodes = None
		self.start = None
		self.success = False
		self.renderer = renderer
		self.expanded_nodes = 0
		self.mode = 0

	def init(self, start):
		self.success = False

		self.start = start
		self.start.hash = start.generateHash()
		self.start.calculateHeuristic()
		
		self.OPEN = []
		self.nodes = {}
		
		self.OPEN.append(start)
		self.nodes[start.hash] =  start
		
		self.render(start)
		
	def setMode(self, mode):
		self.mode = mode
	
	def insertIntoOpen(self, node):
		if(self.mode == 0):
			#BEST-FIRST
			self.OPEN.insert(self.binarySearch(self.OPEN, 0, len(self.OPEN), node), node)
		elif(self.mode == 1):
			#DEPTH-FIRST
			self.OPEN.insert(0, node)
		else:
		#BREADTH-FIRST
			self.OPEN.append(node)
		
		
	def binarySearch(self, list, start, end, node):
		if (start == end):
			return start
		
		middle = int((start + end) / 2)
		if (node.getF() <= list[middle].getF()):
			return self.binarySearch(list, start, middle, node)
		else:
			return self.binarySearch(list, middle + 1, end, node)
		
	
	def attachAndEval(self, parent, child):
		child.h = child.calculateHeuristic()
		child.g = parent.g + parent.childCost.get(child.hash)
		child.parent = parent
		
	def propagate(self, node):
		g = 0.0
		g = node.g
		
		for child in node.children:
			newG = g + node.childCost[child.hash]
			if (newG < child.g):
				child.parent = node
				child.g = newG
				
				if(not child.expanded):
					self.OPEN.remove(child)
					self.insertIntoOpen(child)
					
				self.propagate(child)
			

	def render(self, node):
		if (not self.renderer is None):
			self.renderer.render(self, node)
			return
		
	def step(self):
		#time.sleep(1.0/60.0)
		node = self.OPEN.pop(0)
		#print("Popped node: {}\tHash: {}\tG: {}\tH: {}\tF: {}".format(expandedNodes, node.hash, node.g, node.h, node.getF()))
		#Check if node was goalnode
		if(node.is_goal_node()):
			print("Success!")
			self.success = True
			self.render(node)
			return node
		
		node.expanded = True
		
		#Expand
		children = node.generateChildren()
		
		for i in range(len(children)):
			freshNode = False
			child = children[i]
			
			#Check if childNode exists
			childNode = self.nodes.get(child.hash)
			
			if (childNode is None):
				#node not found yet, make new one.
				childNode = child
				freshNode = True
			
			node.addChild(childNode, child.g - node.g)
			
			if(freshNode):
				self.attachAndEval(node, childNode)
				self.nodes[childNode.hash] = childNode
				self.insertIntoOpen(childNode)

			elif(childNode.g > node.g + node.childCost.get(childNode.hash)):
				self.attachAndEval(node, childNode)
				
				if(childNode.expanded):
					self.propagate(childNode)
				else:
					self.OPEN.remove(childNode)
					self.insertIntoOpen(childNode)	

		self.expanded_nodes += 1		
		self.render(node)
		
	def run(self):
		self.expanded_nodes = 0
		print("Running in mode: {}".format(self.mode))
		if(self.OPEN is None or self.nodes is None or self.start is None or self.success is True):
			print("Algorithm not properly initialized")
			return
		
		while(len(self.OPEN) > 0 and not self.success):
			node = self.step()
		return node